### PR TITLE
2. Refactor Connection & Retry Handling

### PR DESCRIPTION
**Pull request by timoxley**
* Moved majority of connection & retry handling logic into `Connection`.
* Connection is checked and potentially reconnected whenever socket is written to, simplifying everything that writes to socket.
* Improved reconnection/autoconnect logic, should be more predictable now. 
* Manually calling `disconnect()` disables autoconnect. Will only reconnect if asked.
* Cleaned up and improved logging significantly, e.g. now logs the request object/deserialised response rather than hard to read serialized array.
* Error handling also improved, tried to prevent cases of double logging or even double emitting the same error.
* `connect`/`disconnect` now roughly equivalent what `ensureConnected`/`ensureDisconnected` did, methods left for backwards compatibility, could be removed though.

Builds on:
* Split Client into Publisher/Subscriber/Resender https://github.com/streamr-dev/streamr-client-javascript/pull/156

Followed by:
* Publish/Subscribe Promisification https://github.com/streamr-dev/streamr-client-javascript/pull/162
* Publish Refactoring https://github.com/streamr-dev/streamr-client-javascript/pull/164

### CODE REVIEW

**Review by mirpo — APPROVED** [review]
lgtm

File: src/Connection.js [hunk 1]
```diff
@@ -1,169 +1,614 @@
 import EventEmitter from 'eventemitter3'
-import debugFactory from 'debug'
+import Debug from 'debug'
```
[the remark this answers is not in the document]
Comment (timoxley) [reply src/Connection.js:2]: yep `pino` looks good to me, that's on my todo

File: src/Connection.js [hunk 2]
```diff
@@ -1,169 +1,614 @@
...
+            if (!url) {
+                throw new ConnectionError('URL is not defined!')
+            }
+            const socket = new WebSocket(url, ...args)
+            socket.id = uniqueId('socket')
+            socket.binaryType = 'arraybuffer'
+            let opened = 0
+            socket.onopen = () => {
+                opened = 1
+                openSockets += opened
```
[the remark this answers is not in the document]
Comment (timoxley) [reply src/Connection.js:44]: `openSockets -= opened`

Bookkeeping, `onclose` should only decrement if we actually incremented the counter. Could be a bool I guess.

File: src/Connection.js [hunk 3]
```diff
@@ -1,169 +1,614 @@
...
+            try {
+                socket.close()
+            } catch (err) {
+                reject(err)
+                return
+            }
+        }
+
+        if (socket.readyState === WebSocket.CLOSING) {
+            socket.addEventListener('close', resolve)
```
[the remark this answers is not in the document]
Comment (timoxley) [reply src/Connection.js:89]: Was that using `ws` directly or using the old client? `socket.readyState === WebSocket.OPENING` is the connecting state, which is checked above, but does only wait for `'error'`/`'open'` events, assumes these are the only two possible events. Perhaps it should also wait for `'close'` in the case of a close without an error before it hits `'open'`? and perhaps could do with a timeout. Any idea how to trigger that stuck state?

File: src/Connection.js [hunk 4]
```diff
@@ -1,169 +1,614 @@
...
+            this.debug('emit', event)
         }
 
-        if (this.state === Connection.State.DISCONNECTING) {
-            return new Promise((resolve) => {
-                this.once('disconnected', () => resolve(this.connect()))
-            })
+        // note if event handler is async and it rejects we're kinda hosed
+        // until node lands unhandledrejection support
+        // in eventemitter
```
[the remark this answers is not in the document]
Comment (timoxley) [reply src/Connection.js:162]: I'm not sure if that will ever be updated to capture rejections.

Check this: https://nodejs.org/api/events.html#events_capture_rejections_of_promises

https://github.com/nodejs/node/blob/17ebd464ccdf12a4fb46334ff5d7a71f0f2e70a9/lib/events.js#L165-L206

File: src/Connection.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (harbu) [reply src/Connection.js:44]: As this is only for tests, it shouldn't be a big problem.

We ended up with a similar sort of counter not being an accurate representation of the actual number of live connections in a broker node. Seemed like the `disconnect` event didn't always trigger. Apparently if the connection got timed out or smth, the `disconnect` event wouldn't trigger. Ended up using a counter within the library to print the connection count to logs.

**Review by harbu — APPROVED** [review]

**Pull request merged**